# Logistic Regression<br> Treatment Effect

In [1]:
#Import important package
library(Ecdat)
library(ggplot2) 
library(ISLR)
library(tidyverse)
library("ggExtra")
library("ROCR")

Warning message:
"package 'Ecdat' was built under R version 3.4.1"Loading required package: Ecfun
Warning message:
"package 'Ecfun' was built under R version 3.4.4"
Attaching package: 'Ecfun'

The following object is masked from 'package:base':

    sign


Attaching package: 'Ecdat'

The following object is masked from 'package:datasets':

    Orange

Warning message:
"package 'ggplot2' was built under R version 3.4.4"Warning message:
"package 'ISLR' was built under R version 3.4.4"Warning message:
"package 'tidyverse' was built under R version 3.4.4"-- Attaching packages --------------------------------------- tidyverse 1.2.1 --
v tibble  1.4.2     v purrr   0.2.5
v tidyr   0.8.1     v dplyr   0.7.6
v readr   1.1.1     v stringr 1.3.1
v tibble  1.4.2     v forcats 0.3.0
Warning message:
"package 'tibble' was built under R version 3.4.4"Warning message:
"package 'tidyr' was built under R version 3.4.4"Warning message:
"package 'readr' was built under R version 3.4.4"Warning message:
"p

Questions 1-4 are based on the Conceptual question at the end of Chapter 4 in the JWHT textbook (6th edition). The scenario involves data collected about students in a graduate statistics class with variables X1 = hours studied, X2 = undergraduate GPA, and Y = 1 if the student received an A, 0 otherwise. A logistic regression is fitted, and the estimated coefficients are b0 = -6, b1 = 0.05, and b2 = 1.

- ** Estimate the probability that a student who studies for 30 hours and has an undergraduate GPA of 3.5 gets an A in the class. **

In [2]:
b0=-6
b1=0.05
b2=1
X1=30
X2=3.5
#odds for is log(y)
y=exp(b0+b1*X1+b2*X2)
p=y/(1+y)
p

[1] 0.2689414

- ** Estimate the odds ratio in favor of student, who studies for 40 hours and has an undergraduate GPA of 3.5, getting an A in the class. **

In [3]:
b0=-6
b1=0.05
b2=1
X1=40
X2=3.5
odds_for=exp(b0+b1*X1+b2*X2)
odds_for

[1] 0.6065307

- ** Estimate the probability that a student who studies for 40 hours and has an undergraduate GPA of 3.9 gets an A in the class. **

In [4]:
b0=-6
b1=0.05
b2=1
X1=40
X2=3.9
#odds for is log(y)
y=exp(b0+b1*X1+b2*X2)
p=y/(1+y)
p

[1] 0.4750208

- ** How many hours should a student study with an undergraduate GPA of 3.5 to have a probability of getting an A in the class equal to 0.8 **


In [5]:
b0=-6
b1=0.05
b2=1
X2=3.5
p=0.8
log_odds=log(p/(1-p))
X1=(log_odds-b0-b2*X2)/b1
X1

[1] 77.72589

** Question 5-8 are based on the Applied Problem 10 at the end of Chapter 4 in the JWHT textbook. Load the dataset named Weekly (in the ISLR package) into R. Create a new data frame (call is C1) that is a copy of Weekly for your analysis.  Assume “UP” as positive event and “DOWN” as negative event. **

In [6]:
C1<-data.frame(Weekly) %>%
    mutate(resp=ifelse(Direction=="Down",0,1))
head(C1)


Warning message:
"package 'bindrcpp' was built under R version 3.4.4"

Year Lag1   Lag2   Lag3   Lag4   Lag5   Volume    Today  Direction resp
1 1990  0.816  1.572 -3.936 -0.229 -3.484 0.1549760 -0.270 Down      0   
2 1990 -0.270  0.816  1.572 -3.936 -0.229 0.1485740 -2.576 Down      0   
3 1990 -2.576 -0.270  0.816  1.572 -3.936 0.1598375  3.514 Up        1   
4 1990  3.514 -2.576 -0.270  0.816  1.572 0.1616300  0.712 Up        1   
5 1990  0.712  3.514 -2.576 -0.270  0.816 0.1537280  1.178 Up        1   
6 1990  1.178  0.712  3.514 -2.576 -0.270 0.1544440 -1.372 Down      0

- ** Perform a logistic regression with Direction as the response variable and the five lag variables plus Volume as predictors. Use the summary function to review the results.  **

In [7]:
glm.1<- glm(formula=resp~Lag1+Lag2+Lag3+Lag4+Lag5+Volume,data=C1,family="binomial")
summary(glm.1)


Call:
glm(formula = resp ~ Lag1 + Lag2 + Lag3 + Lag4 + Lag5 + Volume, 
    family = "binomial", data = C1)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-1.6949  -1.2565   0.9913   1.0849   1.4579  

Coefficients:
            Estimate Std. Error z value Pr(>|z|)   
(Intercept)  0.26686    0.08593   3.106   0.0019 **
Lag1        -0.04127    0.02641  -1.563   0.1181   
Lag2         0.05844    0.02686   2.175   0.0296 * 
Lag3        -0.01606    0.02666  -0.602   0.5469   
Lag4        -0.02779    0.02646  -1.050   0.2937   
Lag5        -0.01447    0.02638  -0.549   0.5833   
Volume      -0.02274    0.03690  -0.616   0.5377   
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 1496.2  on 1088  degrees of freedom
Residual deviance: 1486.4  on 1082  degrees of freedom
AIC: 1500.4

Number of Fisher Scoring iterations: 4


- ** Do any of the predictors appear to be statistically significant at alpha = 0.05? If so, which one(s)? <br> Lag2 **

 - ** Compute the confusion matrix with 0.5 as the cutoff value. What are the values of true negative, false positive, false negative, true positive, and accuracy? **

In [8]:
C1 <-  C1 %>% 
  mutate(pred_prob_model4 = predict(glm.1, newdata = ., type = "response")) %>% 
  mutate(pred_outcome_model4 = ifelse(pred_prob_model4 >= 0.5,1,0))
#Confusion matrix
xtabs(~resp + pred_outcome_model4, data = C1)

    pred_outcome_model4
resp   0   1
   0  54 430
   1  48 557

In [9]:
TN=54
FP=430
FN=48
TP=557
accuracy=(TN+TP)/(TN+FP+FN+TP)
TN
FP
FN
TP
accuracy

[1] 54

[1] 430

[1] 48

[1] 557

[1] 0.5610652

- ** Fit a logistic regression model using a training data period from 1990 to 2008 with Direction as the response variable and with Lag2 as the only predictor. Compute the confusion matrix and the values of true positive, false negative, true negative, false positive for the holdout data (that is the data from 2009 to 2010). What are the values of true negative, false positive, false negative, true positive, and accuracy? **

In [10]:
C1_subset<-C1 %>% 
    filter(Year>=1990 & Year<=2008)
C1_response<-C1 %>%
    filter(Year>=2009 & Year<=2010)
glm.2<- glm(formula=resp~Lag2,data=C1_subset,family="binomial")
C1_response <-  C1_response %>% 
  mutate(pred_prob_model2 = predict(glm.2, newdata = ., type = "response")) %>% 
  mutate(pred_outcome_model2 = ifelse(pred_prob_model2 >= 0.5,1,0))
#Confusion matrix
xtabs(~resp + pred_outcome_model2, data = C1_response)

    pred_outcome_model2
resp  0  1
   0  9 34
   1  5 56

In [11]:
TN=9
FP=34
FN=5
TP=56
accuracy=(TN+TP)/(TN+FP+FN+TP)
TN
FP
FN
TP
accuracy

[1] 9

[1] 34

[1] 5

[1] 56

[1] 0.625

- ** estimate the Sensitivity and Specificity of the model. **

In [12]:
sensitivity=TP/(TP+FN)
specificity=TN/(TN+FP)
sensitivity
specificity

[1] 0.9180328

[1] 0.2093023

** For the remaining questions (14 and 15) use the star dataset in the ECDAT package. Create a new dataset mydata that only has records for the small and regular.with.aide classes. We are not interested in regular sized classes. Create a dummy variable called small which is 1 if a record in mydata is of a student in a small size class, and is 0 otherwise. Create totalscore which is the sum of the math and the reading scores for each record. **

In [13]:
mydata<-data.frame(Star) %>%
filter(classk=='small.class' | classk=='regular.with.aide') %>%
mutate(small=ifelse(classk=='small.class',1,0)) %>%
mutate(totalscore=tmathssk+treadssk)
head(mydata)


tmathssk treadssk classk            totexpk sex  freelunk race  schidkn small
1 473      447      small.class        7      girl no       white 63      1    
2 536      450      small.class       21      girl no       black 20      1    
3 463      439      regular.with.aide  0      boy  yes      black 19      0    
4 489      447      small.class        5      boy  yes      white 79      1    
5 423      395      regular.with.aide 17      girl yes      black 16      0    
6 439      478      small.class       11      girl no       black 11      1    
  totalscore
1 920       
2 986       
3 902       
4 936       
5 818       
6 917

- ** Please run a linear regression using all the data in mydata using totalscore as the response variable and small as the predictor **

In [14]:
lm.1<-lm(formula=totalscore~small,data=mydata)
summary(lm.1)


Call:
lm(formula = totalscore ~ small, data = mydata)

Residuals:
    Min      1Q  Median      3Q     Max 
-199.50  -51.05   -8.05   42.50  334.50 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  918.497      1.645 558.211  < 2e-16 ***
small         13.554      2.420   5.601 2.28e-08 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 73.86 on 3746 degrees of freedom
Multiple R-squared:  0.008305,	Adjusted R-squared:  0.008041 
F-statistic: 31.37 on 1 and 3746 DF,  p-value: 2.283e-08


- ** What is the estimated coefficient of small? What is its t-value? Is small statistically significant? <br> 13.554,5.601,Yes**


- ** Please run a linear regression using all the data in mydata using totalscore as the response variable and small and teacher experience as the predictors.  **

In [15]:
lm.2<-lm(formula=totalscore~small+totexpk,data=mydata)
summary(lm.2)


Call:
lm(formula = totalscore ~ small + totexpk, data = mydata)

Residuals:
    Min      1Q  Median      3Q     Max 
-201.10  -51.45   -7.73   41.85  340.96 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept) 905.3270     2.6143 346.296  < 2e-16 ***
small        14.6502     2.4127   6.072 1.39e-09 ***
totexpk       1.3429     0.2079   6.460 1.18e-10 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 73.46 on 3745 degrees of freedom
Multiple R-squared:  0.01923,	Adjusted R-squared:  0.01871 
F-statistic: 36.72 on 2 and 3745 DF,  p-value: < 2.2e-16


- **What are the estimated coefficient of small and teacher experience? What are their t-value? Are they statistically significant? **

** Small:14.6502 ,6.072,yes   Teacher:1.3429,6.460,yes **